src/nvim/api/vimscript.c

```diff
@@ -745,3 +746,294 @@ Dictionary nvim_parse_expression(String expr, String flags, Boolean highlight, E
   viml_parser_destroy(&pstate);
   return ret;
 }
+
+/// Translates and returns allocated user function name.
+static String user_function_name(ufunc_T *fp)
+{
+  String name;
+  const size_t name_len = STRLEN(fp->uf_name);
+  if (fp->uf_name[0] == K_SPECIAL) {
+    // Replace [ K_SPECIAL KS_EXTRA KE_SNR ] with "<SNR>"
+    name.size = name_len + 2;
+    name.data = xmalloc(name.size + 1);
+    name.data[0] = '<';
+    name.data[1] = 'S';
+    name.data[2] = 'N';
+    name.data[3] = 'R';
+    name.data[4] = '>';
+    memcpy(name.data + 5, fp->uf_name + 3, name_len - 3);
+    name.data[name.size] = '\0';
+  } else {
+    name.size = name_len;
+    name.data = xmalloc(name.size + 1);
+    memcpy(name.data, fp->uf_name, name_len);
+    name.data[name.size] = '\0';
+  }
+  return name;
+}
+
+/// Takes ownership of name.
+static Dictionary user_function_dict(ufunc_T *fp, String name, bool details, bool lines)
+{
+  Dictionary dict = ARRAY_DICT_INIT;
+  size_t dict_size = 2U + (details ? 8 : 0) + (lines ? 1 : 0);
+  kv_resize(dict, dict_size);
+
+  // Function name
+  PUT(dict, "name", STRING_OBJ(name));
+  PUT(dict, "type", CSTR_TO_OBJ("user"));
+
+  if (details) {
+    // Arguments
+    Array args = ARRAY_DICT_INIT;
+    if (fp->uf_args.ga_len > 0) {
+      kv_resize(args, (size_t)fp->uf_args.ga_len);
+      for (int j = 0; j < fp->uf_args.ga_len; j++) {
+        Dictionary arg = ARRAY_DICT_INIT;
+        PUT(arg, "name", CSTR_TO_OBJ((const char *)FUNCARG(fp, j)));
+        if (j >= fp->uf_args.ga_len - fp->uf_def_args.ga_len) {
+          PUT(arg, "default",
+              CSTR_TO_OBJ(((char **)(fp->uf_def_args.ga_data))
+                          [j - fp->uf_args.ga_len + fp->uf_def_args.ga_len]));
+        }
+        kv_push(args, DICTIONARY_OBJ(arg));
+      }
+    }
+    PUT(dict, "args", ARRAY_OBJ(args));
+    PUT(dict, "varargs", BOOLEAN_OBJ(fp->uf_varargs));
+    // Attributes
+    PUT(dict, "abort", BOOLEAN_OBJ(fp->uf_flags & FC_ABORT));
+    PUT(dict, "range", BOOLEAN_OBJ(fp->uf_flags & FC_RANGE));
+    PUT(dict, "dict", BOOLEAN_OBJ(fp->uf_flags & FC_DICT));
+    PUT(dict, "closure", BOOLEAN_OBJ(fp->uf_flags & FC_CLOSURE));
+    // Script
+    PUT(dict, "sid", INTEGER_OBJ(fp->uf_script_ctx.sc_sid));
+    PUT(dict, "lnum", INTEGER_OBJ(fp->uf_script_ctx.sc_lnum));
+  }
+
+  // Source lines
+  if (lines) {
+    Array array = ARRAY_DICT_INIT;
+    if (fp->uf_lines.ga_len > 0) {
+      kv_resize(array, (size_t)fp->uf_lines.ga_len);
+      for (int j = 0; j < fp->uf_lines.ga_len; j++) {
+        kv_push(array, CSTR_TO_OBJ((const char *)FUNCLINE(fp, j)));
+      }
+    }
+    PUT(dict, "lines", ARRAY_OBJ(array));
+  }
+
+  return dict;
+}
+
+/// Takes ownership of name.
```

Comment:
we have `@param[in]` and `@return [allocated]` conventions, I wonder if we need a `[owned]` or `[freed]` convention, e.g.

```suggestion
/// @param[freed] name
```

Reply:
I think that'd be useful. For me `[owned]` and `[freed]` isn't instantly obvious what it means, but I don't have any better idea at the moment.